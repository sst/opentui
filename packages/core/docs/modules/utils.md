# Utils Module  

The utils module provides utility functions for common operations in OpenTUI, including text attribute creation and other helper functions.

## Overview

This module contains helper functions that simplify common tasks when working with OpenTUI components and rendering.

## Text Attribute Utilities

### createTextAttributes

Create text attributes using a friendly object interface instead of bitwise operations:

```typescript
import { createTextAttributes } from '@opentui/core'

// Create attributes with object syntax
const attributes = createTextAttributes({
  bold: true,
  underline: true,
  italic: false
})

// Equivalent to:
// TextAttributes.BOLD | TextAttributes.UNDERLINE
```

### Usage Examples

```typescript
// No attributes (default)
const plain = createTextAttributes()
// Returns: 0

// Single attribute
const bold = createTextAttributes({ bold: true })
// Returns: 1

// Multiple attributes
const fancy = createTextAttributes({
  bold: true,
  italic: true,
  underline: true
})
// Returns: 13 (1 + 4 + 8)

// All attributes
const all = createTextAttributes({
  bold: true,
  dim: true,
  italic: true,
  underline: true,
  blink: true,
  inverse: true,
  hidden: true,
  strikethrough: true
})
// Returns: 255 (all bits set)
```

### Attribute Options

All options are optional and default to `false`:

```typescript
interface TextAttributeOptions {
  bold?: boolean         // Make text bold
  italic?: boolean       // Make text italic
  underline?: boolean    // Underline text
  dim?: boolean         // Dim/faint text
  blink?: boolean       // Blinking text
  inverse?: boolean     // Swap fg/bg colors
  hidden?: boolean      // Hide text (but preserve space)
  strikethrough?: boolean // Strike through text
}
```

## Integration with Components

### With Text Components

```typescript
import { TextRenderable, createTextAttributes } from '@opentui/core'

const text = new TextRenderable('myText', {
  content: 'Important Message',
  attributes: createTextAttributes({
    bold: true,
    underline: true
  })
})
```

### With Box Components

```typescript
import { BoxRenderable, createTextAttributes } from '@opentui/core'

const box = new BoxRenderable('myBox', {
  title: 'Alert',
  titleAttributes: createTextAttributes({
    bold: true,
    inverse: true
  })
})
```

### Dynamic Attribute Updates

```typescript
class InteractiveText extends TextRenderable {
  private baseAttributes = createTextAttributes({ bold: true })
  private hoverAttributes = createTextAttributes({ 
    bold: true, 
    underline: true,
    inverse: true
  })
  
  onMouseEnter() {
    this.attributes = this.hoverAttributes
  }
  
  onMouseLeave() {
    this.attributes = this.baseAttributes
  }
}
```

## Attribute Manipulation

### Combining with Existing Attributes

```typescript
// Start with some attributes
let attrs = createTextAttributes({ bold: true })

// Add more attributes
attrs |= createTextAttributes({ underline: true })

// Remove attributes
attrs &= ~TextAttributes.BOLD

// Toggle attributes
attrs ^= TextAttributes.ITALIC

// Check for attributes
const isBold = (attrs & TextAttributes.BOLD) !== 0
```

### Attribute Presets

Create reusable attribute presets:

```typescript
const styles = {
  heading: createTextAttributes({ 
    bold: true, 
    underline: true 
  }),
  
  error: createTextAttributes({ 
    bold: true, 
    inverse: true 
  }),
  
  muted: createTextAttributes({ 
    dim: true 
  }),
  
  link: createTextAttributes({ 
    underline: true 
  }),
  
  code: createTextAttributes({ 
    inverse: true 
  })
}

// Use presets
text.attributes = styles.heading
errorText.attributes = styles.error
```

## Performance Considerations

### Caching Attributes

Since `createTextAttributes` performs bitwise operations, cache frequently used combinations:

```typescript
// Good - cache the result
class MyComponent {
  private static readonly HIGHLIGHT_ATTRS = createTextAttributes({
    bold: true,
    inverse: true
  })
  
  highlight() {
    this.attributes = MyComponent.HIGHLIGHT_ATTRS
  }
}

// Avoid - creates new value each time
class MyComponent {
  highlight() {
    this.attributes = createTextAttributes({
      bold: true,
      inverse: true
    }) // Recalculated each call
  }
}
```

## Best Practices

### Semantic Naming

Use descriptive names for attribute combinations:

```typescript
const semanticStyles = {
  primary: createTextAttributes({ bold: true }),
  secondary: createTextAttributes({ dim: true }),
  success: createTextAttributes({ bold: true }),
  warning: createTextAttributes({ bold: true, underline: true }),
  danger: createTextAttributes({ bold: true, inverse: true }),
  info: createTextAttributes({ italic: true }),
  disabled: createTextAttributes({ dim: true, strikethrough: true })
}
```

### Terminal Compatibility

Not all terminals support all attributes:

```typescript
// Most compatible
const basic = createTextAttributes({
  bold: true,      // Widely supported
  underline: true  // Widely supported
})

// Less compatible
const advanced = createTextAttributes({
  italic: true,       // Not all terminals
  blink: true,        // Often disabled
  strikethrough: true // Limited support
})

// Check terminal capabilities
const supportsItalic = process.env.TERM_PROGRAM !== 'Apple_Terminal'
const attrs = createTextAttributes({
  bold: true,
  italic: supportsItalic
})
```

## Examples

### Status Indicators

```typescript
function getStatusAttributes(status: string) {
  switch (status) {
    case 'running':
      return createTextAttributes({ bold: true, blink: true })
    case 'success':
      return createTextAttributes({ bold: true })
    case 'error':
      return createTextAttributes({ bold: true, inverse: true })
    case 'warning':
      return createTextAttributes({ bold: true, underline: true })
    case 'disabled':
      return createTextAttributes({ dim: true, strikethrough: true })
    default:
      return createTextAttributes()
  }
}
```

### Progressive Enhancement

```typescript
// Start with basic styling
let attributes = createTextAttributes({ bold: true })

// Add enhancements based on context
if (isImportant) {
  attributes |= TextAttributes.UNDERLINE
}

if (isError) {
  attributes |= TextAttributes.INVERSE
}

if (isDeprecated) {
  attributes |= TextAttributes.STRIKETHROUGH
}
```

## API Reference

### Functions

- `createTextAttributes(options?: TextAttributeOptions): number`
  - Creates text attributes from an options object
  - Returns a number with appropriate bits set
  - All options default to false

### Types

```typescript
interface TextAttributeOptions {
  bold?: boolean
  italic?: boolean
  underline?: boolean
  dim?: boolean
  blink?: boolean
  inverse?: boolean
  hidden?: boolean
  strikethrough?: boolean
}
```

## Related Modules

- [Types](./types.md) - TextAttributes enum definition
- [Components](./components.md) - Components that use attributes
- [Lib](./lib.md) - Additional text styling utilities
- [Rendering](./rendering.md) - How attributes are rendered