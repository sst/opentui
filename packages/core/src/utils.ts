import { TextAttributes } from "./types"
import { Renderable } from "./Renderable"

export function createTextAttributes({
  bold = false,
  italic = false,
  underline = false,
  dim = false,
  blink = false,
  inverse = false,
  hidden = false,
  strikethrough = false,
}: {
  bold?: boolean
  italic?: boolean
  underline?: boolean
  dim?: boolean
  blink?: boolean
  inverse?: boolean
  hidden?: boolean
  strikethrough?: boolean
} = {}): number {
  let attributes = TextAttributes.NONE

  if (bold) attributes |= TextAttributes.BOLD
  if (italic) attributes |= TextAttributes.ITALIC
  if (underline) attributes |= TextAttributes.UNDERLINE
  if (dim) attributes |= TextAttributes.DIM
  if (blink) attributes |= TextAttributes.BLINK
  if (inverse) attributes |= TextAttributes.INVERSE
  if (hidden) attributes |= TextAttributes.HIDDEN
  if (strikethrough) attributes |= TextAttributes.STRIKETHROUGH

  return attributes
}

// For debugging purposes
export function visualizeRenderableTree(renderable: Renderable, maxDepth: number = 10): void {
  function buildTreeLines(
    node: Renderable,
    prefix: string = "",
    parentPrefix: string = "",
    isLastChild: boolean = true,
    depth: number = 0,
  ): string[] {
    if (depth >= maxDepth) {
      return [`${prefix}${node.id} ... (max depth reached)`]
    }

    const lines: string[] = []
    const children = node.getChildren()

    // Add current node
    lines.push(`${prefix}${node.id}`)

    if (children.length > 0) {
      const lastChildIndex = children.length - 1

      children.forEach((child, index) => {
        const childIsLast = index === lastChildIndex
        const connector = childIsLast ? "└── " : "├── "
        const childPrefix = parentPrefix + (isLastChild ? "    " : "│   ")
        const childLines = buildTreeLines(child, childPrefix + connector, childPrefix, childIsLast, depth + 1)
        lines.push(...childLines)
      })
    }

    return lines
  }

  const treeLines = buildTreeLines(renderable)
  console.log("Renderable Tree:\n" + treeLines.join("\n"))
}
