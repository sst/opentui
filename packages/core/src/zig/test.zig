const std = @import("std");

// Import all test modules
const text_buffer_tests = @import("tests/text-buffer_test.zig");
const text_buffer_highlights_tests = @import("tests/text-buffer-highlights_test.zig");
const text_buffer_view_tests = @import("tests/text-buffer-view_test.zig");
const text_buffer_selection_tests = @import("tests/text-buffer-selection_test.zig");
const text_buffer_drawing_tests = @import("tests/text-buffer-drawing_test.zig");
const text_buffer_segment_tests = @import("tests/text-buffer-segment_test.zig");
const text_buffer_iterators_tests = @import("tests/text-buffer-iterators_test.zig");
const edit_buffer_tests = @import("tests/edit-buffer_test.zig");
const edit_buffer_history_tests = @import("tests/edit-buffer-history_test.zig");
const editor_view_tests = @import("tests/editor-view_test.zig");
const grapheme_tests = @import("tests/grapheme_test.zig");
const syntax_style_tests = @import("tests/syntax-style_test.zig");
const rope_tests = @import("tests/rope_test.zig");
const rope_nested_tests = @import("tests/rope-nested_test.zig");
const rope_fuzz_tests = @import("tests/rope_fuzz_test.zig");
const utf8_tests = @import("tests/utf8_test.zig");
const event_emitter_tests = @import("tests/event-emitter_test.zig");
const buffer_tests = @import("tests/buffer_test.zig");
// const example_tests = @import("example_test.zig");

// Re-export test declarations from individual test files
// This allows `zig test index.zig` to run all tests
comptime {
    _ = text_buffer_tests;
    _ = text_buffer_highlights_tests;
    _ = text_buffer_view_tests;
    _ = text_buffer_selection_tests;
    _ = text_buffer_drawing_tests;
    _ = text_buffer_segment_tests;
    _ = text_buffer_iterators_tests;
    _ = edit_buffer_tests;
    _ = edit_buffer_history_tests;
    _ = editor_view_tests;
    _ = grapheme_tests;
    _ = syntax_style_tests;
    _ = rope_tests;
    _ = rope_nested_tests;
    _ = rope_fuzz_tests;
    _ = utf8_tests;
    _ = event_emitter_tests;
    _ = buffer_tests;
    // _ = example_tests;
}
