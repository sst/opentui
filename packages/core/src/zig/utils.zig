const std = @import("std");

/// RGBA color type (4 f32 values)
pub const RGBA = @Vector(4, f32);

/// Convert a pointer to 4 f32 values into an RGBA color
pub fn f32PtrToRGBA(ptr: [*]const f32) RGBA {
    return .{ ptr[0], ptr[1], ptr[2], ptr[3] };
}
