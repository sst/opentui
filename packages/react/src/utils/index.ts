import {
  InputRenderable,
  InputRenderableEvents,
  isRenderable,
  SelectRenderable,
  SelectRenderableEvents,
  TabSelectRenderable,
  TabSelectRenderableEvents,
} from "@opentui/core"
import type { Instance, Props, Type } from "../types/host"

function initEventListeners(instance: Instance, eventName: string, listener: any, previousListener?: any) {
  if (previousListener) {
    instance.off(eventName, previousListener)
  }

  if (listener) {
    instance.on(eventName, listener)
  }
}

function setStyle(instance: Instance, styles: any, oldStyles: any) {
  if (styles && typeof styles === "object") {
    if (oldStyles != null) {
      for (const styleName in styles) {
        const value = styles[styleName]
        if (styles.hasOwnProperty(styleName) && oldStyles[styleName] !== value) {
          // @ts-expect-error props are not strongly typed in the reconciler, so we need to allow dynamic property access
          instance[styleName] = value
        }
      }
    } else {
      for (const styleName in styles) {
        if (styles.hasOwnProperty(styleName)) {
          const value = styles[styleName]
          // @ts-expect-error props are not strongly typed in the reconciler, so we need to allow dynamic property access
          instance[styleName] = value
        }
      }
    }
  }
}

function setProperty(instance: Instance, type: Type, propKey: string, propValue: any, oldPropValue?: any) {
  switch (propKey) {
    case "onChange":
      if (instance instanceof InputRenderable) {
        initEventListeners(instance, InputRenderableEvents.CHANGE, propValue, oldPropValue)
      } else if (instance instanceof SelectRenderable) {
        initEventListeners(instance, SelectRenderableEvents.SELECTION_CHANGED, propValue, oldPropValue)
      } else if (instance instanceof TabSelectRenderable) {
        initEventListeners(instance, TabSelectRenderableEvents.SELECTION_CHANGED, propValue, oldPropValue)
      }
      break
    case "onInput":
      if (instance instanceof InputRenderable) {
        initEventListeners(instance, InputRenderableEvents.INPUT, propValue, oldPropValue)
      }
      break
    case "onSubmit":
      if (instance instanceof InputRenderable) {
        initEventListeners(instance, InputRenderableEvents.ENTER, propValue, oldPropValue)
      }
      break
    case "onSelect":
      if (instance instanceof SelectRenderable) {
        initEventListeners(instance, SelectRenderableEvents.ITEM_SELECTED, propValue, oldPropValue)
      } else if (instance instanceof TabSelectRenderable) {
        initEventListeners(instance, TabSelectRenderableEvents.ITEM_SELECTED, propValue, oldPropValue)
      }
      break
    case "focused":
      if (isRenderable(instance)) {
        if (!!propValue) {
          instance.focus()
        } else {
          instance.blur()
        }
      }
      break
    case "style":
      setStyle(instance, propValue, oldPropValue)
      break
    case "children":
      // Skip children handling - React reconciler handles this automatically
      break
    default:
      // @ts-expect-error props are not strongly typed in the reconciler, so we need to allow dynamic property access
      instance[propKey] = propValue
  }
}

export function setInitialProperties(instance: Instance, type: Type, props: Props) {
  for (const propKey in props) {
    if (!props.hasOwnProperty(propKey)) {
      continue
    }

    const propValue = props[propKey]
    if (propValue == null) {
      continue
    }

    setProperty(instance, type, propKey, propValue)
  }
}

export function updateProperties(instance: Instance, type: Type, oldProps: Props, newProps: Props) {
  for (const propKey in oldProps) {
    const oldProp = oldProps[propKey]
    if (oldProps.hasOwnProperty(propKey) && oldProp != null && !newProps.hasOwnProperty(propKey)) {
      setProperty(instance, type, propKey, null, oldProp)
    }
  }

  for (const propKey in newProps) {
    const newProp = newProps[propKey]
    const oldProp = oldProps[propKey]

    if (newProps.hasOwnProperty(propKey) && newProp !== oldProp && (newProp != null || oldProp != null)) {
      setProperty(instance, type, propKey, newProp, oldProp)
    }
  }
}
