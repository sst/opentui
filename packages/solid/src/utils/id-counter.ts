const idCounter = new Map<string, number>();

export function getNextId(elementType: string): string {
  if (!idCounter.has(elementType)) {
    idCounter.set(elementType, 0);
  }

  const value = idCounter.get(elementType)! + 1;
  idCounter.set(elementType, value);
  return `${elementType}-${value}`;
}
