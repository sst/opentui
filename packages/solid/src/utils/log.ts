export const log = (...args: any[]) => {
  if (process.env.DEBUG) {
    console.log("[Reconciler]", ...args)
  }
}
