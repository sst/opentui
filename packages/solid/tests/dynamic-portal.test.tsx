import { describe, expect, it, beforeEach, afterEach } from "bun:test"
import { testRender, Dynamic, Portal } from "../index"
import { createSignal, Show } from "solid-js"
import { createSpy } from "./utils/spy"
import type { BoxRenderable } from "@opentui/core"

let testSetup: Awaited<ReturnType<typeof testRender>>

describe("SolidJS Renderer - Dynamic and Portal Components", () => {
  beforeEach(async () => {
    if (testSetup) {
      testSetup.renderer.destroy()
    }
  })

  afterEach(() => {
    if (testSetup) {
      testSetup.renderer.destroy()
    }
  })

  describe("<Dynamic> Component", () => {
    it("should handle undefined component gracefully", async () => {
      testSetup = await testRender(
        () => (
          <box>
            <Dynamic component={undefined}>This should not render</Dynamic>
          </box>
        ),
        { width: 20, height: 5 },
      )

      await testSetup.renderOnce()
      const children = testSetup.renderer.root.getChildren()[0]!.getChildren()
      expect(children.length).toBe(0)
    })

    it("should pass props correctly to dynamic components", async () => {
      const [color, setColor] = createSignal("red")
      const [text, setText] = createSignal("Initial text")

      testSetup = await testRender(
        () => (
          <box>
            <Dynamic component="text" style={{ fg: color() }}>
              {text()}
            </Dynamic>
          </box>
        ),
        { width: 20, height: 3 },
      )

      await testSetup.renderOnce()
      let frame = testSetup.captureCharFrame()
      expect(frame).toContain("Initial text")

      setColor("blue")
      setText("Updated text")
      await testSetup.renderOnce()
      frame = testSetup.captureCharFrame()
      expect(frame).toContain("Updated text")
      expect(frame).toMatchSnapshot()
    })

    it("should handle event handlers in dynamic components", async () => {
      const onInputSpy = createSpy()

      testSetup = await testRender(
        () => (
          <box>
            <Dynamic component="input" focused={true} onInput={onInputSpy} />
          </box>
        ),
        { width: 20, height: 5 },
      )

      await testSetup.mockInput.typeText("test")

      expect(onInputSpy.callCount()).toBe(4)
      expect(onInputSpy.calls[0]?.[0]).toBe("t")
      expect(onInputSpy.calls[3]?.[0]).toBe("test")
    })

    it("should handle false Show inside dynamic that switches between text and box", async () => {
      /* Tests for slot renderable being able handle switching between a LayoutSlot and a TextSlot
       * Expected to just run without crash
       */
      const [componentType, setComponentType] = createSignal<"text" | "box">("text")

      testSetup = await testRender(
        () => (
          <box>
            <Dynamic component={componentType()}>
              <Show when={false}>This should never render</Show>
            </Dynamic>
          </box>
        ),
        { width: 20, height: 5 },
      )

      await testSetup.renderOnce()

      setComponentType("box")
      await testSetup.renderOnce()
    })
  })

  describe("<Portal> Component", () => {
    it("should render content to default mount point", async () => {
      testSetup = await testRender(
        () => (
          <box>
            <text>Before portal</text>
            <Portal>
              <text>Portal content</text>
            </Portal>
            <text>After portal</text>
          </box>
        ),
        { width: 25, height: 8 },
      )

      await testSetup.renderOnce()
      const frame = testSetup.captureCharFrame()
      expect(frame).toContain("Portal content")
    })

    it("should render content to custom mount point", async () => {
      let customMount!: BoxRenderable

      testSetup = await testRender(
        () => (
          <box>
            <box ref={customMount} />
            <Portal mount={customMount}>
              <box style={{ border: true }} title="Portal Box">
                <text>Portal content</text>
              </box>
            </Portal>
          </box>
        ),
        { width: 25, height: 8 },
      )

      await testSetup.renderOnce()
      const frame = testSetup.captureCharFrame()
      expect(frame).toContain("Portal content")
      expect(customMount.getChildren().length).toBe(1)
    })

    it("should handle complex nested content in portal", async () => {
      testSetup = await testRender(
        () => (
          <box>
            <Portal>
              <text>Nested text 1</text>
              <text>Nested text 2</text>
            </Portal>
          </box>
        ),
        { width: 30, height: 10 },
      )

      await testSetup.renderOnce()
      const frame = testSetup.captureCharFrame()
      expect(frame).toContain("Nested text 1")
      expect(frame).toContain("Nested text 2")
    })

    it("should handle portal cleanup on unmount", async () => {
      const [showPortal, setShowPortal] = createSignal(true)

      testSetup = await testRender(
        () => (
          <box>
            <Show when={showPortal()}>
              <Portal>
                <text>Portal content</text>
              </Portal>
            </Show>
          </box>
        ),
        { width: 20, height: 5 },
      )

      await testSetup.renderOnce()
      let frame = testSetup.captureCharFrame()
      expect(frame).toContain("Portal content")

      setShowPortal(false)
      await testSetup.renderOnce()
      frame = testSetup.captureCharFrame()
      expect(frame).not.toContain("Portal content")
    })

    it("should handle multiple portals", async () => {
      testSetup = await testRender(
        () => (
          <box>
            <Portal>
              <text>First portal</text>
            </Portal>
            <Portal>
              <text>Second portal</text>
            </Portal>
          </box>
        ),
        { width: 25, height: 8 },
      )

      await testSetup.renderOnce()
      const frame = testSetup.captureCharFrame()
      expect(frame).toContain("First portal")
      expect(frame).toContain("Second portal")
      expect(testSetup.renderer.root.getChildren().length).toBe(3)
    })
  })

  describe("<Dynamic> + <Portal> Integration", () => {
    it("should handle Dynamic component inside Portal", async () => {
      const [useComponentA, setUseComponentA] = createSignal(true)

      testSetup = await testRender(
        () => {
          const ComponentA = () => <text>Portal Component A</text>
          const ComponentB = () => <text>Portal Component B</text>

          return (
            <box>
              <Portal>
                <Dynamic component={useComponentA() ? ComponentA : ComponentB} />
              </Portal>
            </box>
          )
        },
        { width: 25, height: 8 },
      )

      await testSetup.renderOnce()
      let frame = testSetup.captureCharFrame()
      expect(frame).toContain("Portal Component A")

      setUseComponentA(false)
      await testSetup.renderOnce()
      frame = testSetup.captureCharFrame()
      expect(frame).toContain("Portal Component B")
    })

    it("should handle Portal with Dynamic mount point", async () => {
      const [useCustomMount, setUseCustomMount] = createSignal(false)

      let ref!: BoxRenderable

      testSetup = await testRender(
        () => (
          <box>
            <box ref={ref}>
              <text>Custom target</text>
            </box>
            <Portal mount={useCustomMount() ? ref : undefined}>
              <text>Dynamic mount content</text>
            </Portal>
            <text>Static content</text>
          </box>
        ),
        { width: 30, height: 10 },
      )

      await testSetup.renderOnce()
      let frame = testSetup.captureCharFrame()
      expect(frame).toContain("Dynamic mount content")
      expect(ref.getChildren().length).toBe(1)

      setUseCustomMount(true)
      await testSetup.renderOnce()
      frame = testSetup.captureCharFrame()
      expect(frame).toContain("Dynamic mount content")
      expect(ref.getChildren().length).toBe(2)

      setUseCustomMount(false)
      frame = testSetup.captureCharFrame()
      expect(frame).toContain("Dynamic mount content")
      expect(ref.getChildren().length).toBe(1)
    })

    it("should handle switching between Portal and non-Portal with Dynamic", async () => {
      const [usePortal, setUsePortal] = createSignal(true)

      let ref!: BoxRenderable

      testSetup = await testRender(
        () => (
          <box ref={ref}>
            <Dynamic component={usePortal() ? Portal : "box"} {...(usePortal() ? {} : { style: { border: true } })}>
              <text>Conditional portal content</text>
            </Dynamic>
          </box>
        ),
        { width: 30, height: 8 },
      )

      await testSetup.renderOnce()
      let frame = testSetup.captureCharFrame()
      expect(frame).toContain("Conditional portal content")
      expect(testSetup.renderer.root.getChildren().length).toBe(2)

      setUsePortal(false)
      await testSetup.renderOnce()
      frame = testSetup.captureCharFrame()
      expect(frame).toContain("Conditional portal content")
      expect(testSetup.renderer.root.getChildren().length).toBe(1)
    })
  })
})
