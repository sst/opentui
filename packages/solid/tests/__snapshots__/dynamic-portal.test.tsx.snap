// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`SolidJS Renderer - Dynamic and Portal Components <Dynamic> Component should pass props correctly to dynamic components 1`] = `
"Updated text        
                    
                    
"
`;

exports[`SolidJS Renderer - Dynamic and Portal Components <Dynamic> + <Portal> Integration should handle Portal with Dynamic mount point 1`] = `
"Custom target                 
Dynamic mount content         
Static content                
                              
                              
                              
                              
                              
                              
                              
"
`;
